File: 31-next-permutation/31-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        j = len(nums)-2
        for i in range(len(nums)-1,0,-1):
            if(nums[i]>nums[j]):
                break
            j-=1
        if(j==-1):
            nums[:] = nums[::-1]
        else:
            for i in range(len(nums)-1,j,-1):
                if(nums[i]>nums[j]):
                    nums[i],nums[j] = nums[j], nums[i]
                    break
            nums[j+1:] = nums[-1:j:-1]
        
            
        

File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        int n = nums.size();
        int i = n-2;
        while(i >= 0){
            if(nums[i] < nums[i+1]) break;
            i--;
        }
        if(i==-1){
            reverse(nums.begin(), nums.end());
            return;
        }
        int j = i+1;
        int mini = INT_MAX;
        int index;
        while(j < n){
            if(nums[j] > nums[i]){
                mini = min(mini, nums[j]);
                index = j;
            }
            j++;
        }
        swap(nums[i], nums[index]);
        reverse(nums.begin()+i+1, nums.end());
        return;
    }
};